WIP Refactor VivViewer

ilan-gold opened the pull request:
This is....a big one.  I'm requesting comment because I don't know exactly how this will fit in with Vitessce.  I've tried side-by-side viewing in the Viv demo and it works very well.  At a high level this PR:

- Makes `VivViewer` accept a list of `views` (subclasses of `VivView`), `viewStates`, and `layerProps` (props for the layers returned by instances of `VivView`) so we can have different configurations (side-by-side with same or different picture, picture-in-picture etc.).
- Adds documentation from `documentationjs` like http://chgw.gehlenborglab.org/docs/

A few things I think you all can comment on:

- What should the prop given to `VivViewer` for `viewStates` be called?  It has always been prefixed by `initial` but it also communicates height/width, which can update (and needs to).
- Does this API make sense?  Should `layerProps`/`viewState` exist entirely within the views?

Once this is agreed upon at a high level (i.e the "Draft" part of the PR is done) I will:

- Implement helper components like `PictureInPictureViewer`, `SingleImageViewer`, and `SideBySideViewer` inheriting from `VivViewer` that simply take in props (i.e without instantiating some subclass of `VivView`) and manage the classes/state/layers for you.
- Write comprehensive tests based on the agreed upon API

Does this sound good?  I think @manzt is the primary stakeholder here, but feel free to comment at however high a level you want @mccalluc.

Fixes #107 and fixes #111.

ilan-gold commented on src/DetailView.js:
```diff
@@ -0,0 +1,16 @@
+    const { loader } = props;
+    const { id } = this;
+    return new VivViewerLayer(props, {
+      id: `${loader.type}-${id}#`,
```
If one id contains another (like `detail` and `detail2`, we need a way to mark the end of the string.  This was my solution since there are only 2 parts of the `id` at this level.

mccalluc replied:
Cool: That scheme seems fine, I think it just needs to be wrapped so the details aren't exposed.

ilan-gold commented:
Ok @mccalluc, let me see what @manzt has to say, and let me incorporate that with what you're saying, and then I'll push to s3 and re-request your view.

ilan-gold commented on src/DetailView.js:
```diff
@@ -0,0 +1,16 @@
+import VivView from './VivView';
+
+/**
+ * This class generates a VivViewerlayer and a view for use in the VivViewer as a detailed view
```
Yes, the later, at `getLayer`

ilan-gold commented on src/VivViewer.js:
```diff
@@ -1,157 +1,117 @@
   // eslint-disable-next-line class-methods-use-this
   layerFilter({ layer, viewport }) {
-    if (layer.id.includes(viewport.id)) {
+    if (layer.id.includes(`-${viewport.id}#`)) {
```
That makes sense.  Will do.

manzt commented on src/OverviewView.js:
```diff
@@ -0,0 +1,141 @@
+ * This class generates a OverviewLayer and a view for use in the VivViewer as an overview to a Detailview (they must be used in conjection)
+ * @param {Object} viewState The viewState object.
+ * @param {Object} loader The loader, used for inferring zoom level.
+ * @param {number} detailHeight The height of the detail view.
+ * @param {number} detailWidth The width of the detail view.
+ * @param {number} overviewScale The scale of this viewport relative to the detail.
+ * @param {number} margin The margin to be offset from the the corner of the other viewport.
+ * @param {string} overviewLocation The location of the viewport - one of "bottom-right", "top-right", "top-left", "bottom-left."
```
For this class, I'd recommend removing `overview` from the property names since they are redundant of the class name. (`overviewScale` --> `scale`, etc). I'm also somewhat inclined to use `position` rather than `location` (I think I've seen this used elsewhere but I'm actually not sure)...

ilan-gold replied:
Ok.

manzt commented on src/OverviewView.js:
```diff
@@ -0,0 +1,141 @@
+  _setHeightWidthScale({ detailWidth, overviewScale }) {
+    const { loader } = this;
+    const { numLevels } = loader;
+    const rasterSize = loader.getRasterSize({
```
Perhaps the loader method isn't the greatest name, but I'd read this easier as .. 
```javascript
const { width: rasterWidth, height: rasterHeight } = loader.getRasterSize({
```

ilan-gold replied:
Ok.

manzt commented on src/OverviewView.js:
```diff
@@ -0,0 +1,141 @@
+    this._imageHeight = rasterSize.height;
+  }
+
+  /**
+   * Set the x and y of the current view.
+   */
```
```javascript
  /**
   * Set the x and y (top left corner) of this overview relative to the detail.
   */
```

manzt commented on src/OverviewView.js:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  getViewState(viewState) {
+    // Scale the view as the overviewScale changes with screen resizing - basically, do not react to any view state changes.
```
One thing that is occurring to me now (and sorry for bringing this up) is that with using `scale`, `detailWidth`, and `detailHeight` means that if the user adjusts the size of their screen, the size of the detail view will grow and shrink. This is kind of awkward, especially if I have a detail view full screen with a overview view, if I go to half screen the overview would become tiny. Initially I thought having this resize be dynamic would be a nice feature, but I'm almost more inclined to say we should force the user to provide an absolute width (and use an absolute default as well).

ilan-gold replied:
I am a little confused at what you're asking.  It seems you're saying that it's awkward for the `DetailView` to change size, and then by extension, for the `OverviewView` to change with it.  But the detail view doesn't have to resize - that is in `App.js` and is not a built-in part of `Viv`, and those size parameters are absolute.  We could just not pass in dynamically the changing height and width to `VivViewer`.  Or are you saying that `DetailView` should change size, and `OverviewView` should not?  Or, should `DetailView` not change size, and `OverviewView` remain pinned to whatever corner is in the bottom of the screen?  Sorry for all the questions but I want to understand what you're saying and also get this right.

manzt commented on src/OverviewView.js:
```diff
@@ -0,0 +1,141 @@
+  getLayer({ viewStates, props }) {
+    const { id, overviewScale, loader } = this;
+    // Scale the bounding box.
+    const boundingBox = VivView.makeBoundingBox(viewStates.detail).map(coords =>
```
Kind of odd you are using a static method from the inherited class? I would just make this a `util` function, or because of in definition couldn't you just do `this.makeBoundingBox` ?

ilan-gold replied:
Sure, I can change this.

manzt commented on src/VivView.js:
```diff
@@ -0,0 +1,77 @@
+ * This class generates a layer and a view for use in the VivViewer
+ * @param {number} viewState The viewState object
+ * @param {string} id The id for the current view
+ * @param {string} x The x location on the screen for the current view
+ * @param {string} y The y location on the screen for the current view
```
again I think it's useful to clarify that x and y are top-left corner of view (unless you think this is clear)

manzt commented on src/VivView.js:
```diff
@@ -0,0 +1,77 @@
+   * @param {viewState} Object The viewState for a certain viewport
+   * @returns {View} The DeckGL View for this viewport
+   */
+  static makeBoundingBox(viewState) {
+    const viewport = new OrthographicView().makeViewport({
+      // From the current `detail` viewState, we need its projection matrix (actually the inverse).
+      viewState,
+      height: viewState.height,
+      width: viewState.width
+    });
+    // Use the inverse of the projection matrix to map screen to the view space.
+    return [
+      viewport.unproject([0, 0]),
+      viewport.unproject([viewport.width, 0]),
+      viewport.unproject([viewport.width, viewport.height]),
+      viewport.unproject([0, viewport.height])
+    ];
+  }
```
Again somewhat incline to have this be a `util`. I generally prefer static methods for offering convenience methods to end user (i.e. creating class instances `VivView.fromConfig(configObj) --> new VivView` or if you think people would be inclined to use:

```javascript
import { VivView } from 'viv';

const boundingBox = VivView.makeBoundingBox(state);
```

Or if you foresee others creating inherited classes. In that case keep, but make sure to use `this.makeBoundingBox` on the children otherwise its the same as importing a function..

ilan-gold replied:
Ok.

manzt commented on src/VivViewer.js:
```diff
@@ -1,157 +1,117 @@
+      viewStates: {}
     };
+    const { viewStates } = this.state;
+    const { views } = this.props;
```
```javascript
     const { views, viewStates: propsViewStates } = this.props;
```

ilan-gold replied:
Ok.

manzt commented on src/VivView.js:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Create a viewState for this class, checking the id to make sure this class and veiwState match.
+   * @param {ViewState} Object A viewState object.
+   * @returns {ViewState} The viewState for this class.
```
or null?

manzt commented on src/VivViewer.js:
```diff
@@ -1,157 +1,117 @@
+    const { views } = this.props;
+    views.forEach((view, i) => {
+      // eslint-disable-next-line react/destructuring-assignment
+      viewStates[view.id] = view.getViewState(this.props.viewStates[i]);
```
What does the `getViewState` function do? How is `this.props.viewStates` different from `this.state.viewStates`? Does `getViewState` make some conversion?

ilan-gold replied:
Yes, for example, the `OverviewView` holds various properties constant.  For example, you could see linked zoom-pan for side-by-side behaving differently according to flags for coupling and decoupling the zoom/pan.

manzt commented on src/VivViewer.js:
```diff
@@ -1,157 +1,117 @@
-    }
+    // const { overview } = this.state;
+    const { views } = this.props;
+    this.setState(prevState => {
+      const newState = { ...prevState };
+      views.forEach(view => {
+        newState.viewStates = {
+          ...newState.viewStates,
+          [view.id]:
+            view.getViewState({ ...viewState, id: viewId }) ||
+            prevState.viewStates[view.id]
+        };
+      });
```
Again getting a little lost in the viewStates for the `VivViews` and viewStates in `VivViewer`

manzt commented on src/VivViewer.js:
```diff
@@ -1,157 +1,117 @@
-        overview: overview ? vivOverview.getViewState(initialViewState) : null
-      },
-      overview: vivOverview
+      viewStates: {}
```
Shouldn't initial state be an array?

manzt commented on demo/src/App.js:
```diff
@@ -148,38 +147,51 @@ function App() {
+    ),
+    colormap: colormapOn
+  };
+  const detail = new DetailView({ viewState });
+  const views = [detail];
+  const layerProps = [props];
+  const viewStates = [viewState];
```
This feels overly complicated to me and raises some serious concerns about how `viewState` is managed. Both `views` and `viewStates` are given to the `VivViewer` and both are maintaining the `viewState` somehow.

ilan-gold commented:
@mccalluc  and @manzt Would y'all be willing to re-review and comment?  Demo is here: https://vitessce-data.s3.us-east-1.amazonaws.com/demos/viv/refactor_vivviewer/index.html and documentation is here: https://vitessce-data.s3.us-east-1.amazonaws.com/demos/viv/refactor_vivviewer/docs/index.html.  Thanks!

ilan-gold commented:
@manzt I found `useEffect` to be slower than my `componentDidUpdate` for resizing events.  I think the reason was that I am passing in height/width within the object in `views` and this nested structure doesn't play well.  I would be open to other suggestions - the code is here if you want to look at it side-by-side with the class component https://github.com/hubmapconsortium/vitessce-image-viewer/blob/56c809dab056168232db1c4c0ba0fd404b4cbb7d/src/viewers/VivViewer.js

manzt commented on src/viewers/VivViewer.js:
```diff
@@ -0,0 +1,164 @@
+    const { views, randomize } = this.props;
+    const { viewStates } = this.state;
+    const deckGLViews = views.map(view => view.getDeckGlView());
+    // DeckGL seems to use the first view more than the second for updates
+    // so this forces it to use the others more evenly.  This isn't perfect,
+    // but I am not sure what else to do.  The DeckGL render hooks don't help,
+    // but maybe useEffect() would help?  I couldn't work it out as
+    // The issue is that I'm not sure how React would distinguish between forced updates
+    // from permuting the views array and "real" updates like zoom/pan.
+    // I tried keeping a counter but I couldn't figure out resetting it
+    // without triggering a re-render.
```
I would look into `React.memo`. https://blog.bitsrc.io/improve-performance-in-react-functional-components-using-react-memo-b2e80c11e15a. 

Effectively you memoize the return of a functional component, but also supply a comparator which compares prevProps and nextProps.

```javascript
const App = (props) => { /* ...some expensive logic */ return <div>Calculated expensive thing</div> } 
const shouldUpdate = (prevProps, nextProps) => {
  return prevProps.some.nested.thing !== nextProps.some.nested.thing;
}
export default React.memo(App, shouldUpdate);
```

ilan-gold replied:
Hmmm...Ok, will do.

ilan-gold replied:
See my other comment.  Feel free to respond!

ilan-gold commented:
@mccalluc I will add some documentation as to how this all fits together.  I can also change the name of the `LinkedDetailViewer` if @manzt agrees.  I couldn't come up with anything good, but `SideBySideViewer` is much more human readable, so I'd take it.  

As for the code snippets, I think it's ok to keep them even if they are hard to maintain - I have seen other libraries do this in their docs (https://deck.gl/#/documentation/overview/upgrade-guide?section=tile3dlayer and http://guido.io/zarr.js/#/getting-started/nested-arrays?id=creating-nestedarrays are two examples I can think of) and I find it extremely helpful for getting started.  Even if they grow a bit out of date, the props (`colormap`, `sliderValues` etc.) API has remained largely unchanged even as components have been added and I think it is helpful for people to see everything declared all at once.  That being side I can clean it up a bit.

manzt commented on src/views/DetailView.js:
```diff
@@ -0,0 +1,23 @@
+    const { id } = this;
+    const layer = loader.isPyramid
+      ? new VivViewerLayer(props, {
+          id: loader.type + getVivId(id),
```
Usually I prefer template literals for this kind of thing  (\`${loader.type}${getVivId(id)}\`) simply because it's clearer to me that result is a string.

ilan-gold replied:
Ok.

manzt commented on src/views/VivView.js:
```diff
@@ -0,0 +1,57 @@
+   * @param {ViewState} Object A viewState object.
+   * @returns {ViewState} The viewState for this class (or null by default if the ids do not match).
+   */
+  // eslint-disable-next-line class-methods-use-this
```
still necessary eslint disable?

ilan-gold commented on src/viewers/VivViewer.js:
```diff
@@ -0,0 +1,164 @@
+   */
+  // eslint-disable-next-line class-methods-use-this
+  layerFilter({ layer, viewport }) {
+    return layer.id.includes(getVivId(viewport.id));
```
Sadly, it cannot.  The layer id's are unique and therefore may only contain the id, but not be equal to it.

ilan-gold commented on src/viewers/VivViewer.js:
```diff
@@ -0,0 +1,164 @@
+    navigator.vendor &&
+    navigator.vendor.indexOf('Apple') > -1 &&
+    navigator.userAgent &&
+    navigator.userAgent.indexOf('CriOS') === -1 &&
+    navigator.userAgent.indexOf('FxiOS') === -1
```
I don't think so.  I tested the website last night on mobile and even though it doesn't work, it didn't throw the error.

ilan-gold commented on demo/src/App.js:
```diff
@@ -172,13 +195,45 @@ function App() {
+            </Button>
+          ) : (
+            <>
+              <Button
+                disabled={!isPyramid || isLoading}
+                onClick={toggleZoomLock}
+                variant="outlined"
+                size="small"
+                fullWidth
+              >
+                {zoomLock ? 'Unlock' : 'Lock'} Zoom
+              </Button>
+              <Button
+                disabled={!isPyramid || isLoading}
+                onClick={togglePanLock}
+                variant="outlined"
+                size="small"
+                fullWidth
+              >
+                {panLock ? 'Unlock' : 'Lock'} Pan
+              </Button>
```
I actually liked the no-padding.  I had noticed it as well but left it because it gives a bit of distinction between toggling demos and then controlling within the demo you are on.

ilan-gold commented on src/viewers/VivViewer.js:
```diff
@@ -0,0 +1,164 @@
+            prevProps.views[i].initialViewState.width
+      )
+    ) {
+      // eslint-disable-next-line react/no-did-update-set-state
```
I got `getDerivedStateFromProps` working (I had tried `shouldComponentUpdate` but it wasn't right for this) - thanks for the suggestion!  I found that the functional component works but was somewhat slower updating the `deck.gl` stuff for resizing, and `React.memo` didn't really help because so many different things can change (sliders, colors, colormaps etc.).  None of the `deck.gl` demos use them and I wonder if there is a reason....

In any case, I'm going to file an issue - this is a nice first pass, but I think someone should investigate relative sizing the `DeckGL` component with our use cases.  I know it is in theory possible with `deck.gl` but we do a lot of custom stuff that might make it infeasible/nasty.  I wonder if this change to how sizing is performed would help the functional component performance.

manzt replied:
Great! Generally if I'm writing things in React, it's my preference now to use functional components, but for more complex state I've run into cases where it's easier to use a class and specific lifecycle methods. The concern here was not using the correct lifecycle method and ignoring eslint rules. We certainly can go over this again and refactor, but this seems like a totally appropriate solution. Thanks.

manzt replied:
FWIW, in the end I don't think the trade-offs between class and functional components are usually performance based. Spending enough time, there is usually a way to achieve similar effects. I prefer functional simply because they are easier to read (IMO) and generally result in less code (so easier to maintain).

ilan-gold commented on demo/src/App.js:
```diff
@@ -172,13 +195,45 @@ function App() {
             fullWidth
           >
-            {overviewOn ? 'Hide' : 'Show'} Picture-In-Picture
+            {useLinkedView ? 'Overview' : 'Linked'}
```
Ok, how does this work?

App.js:
`Hide/Show Picture-in-Picture` --> `Hide/Show Overview`
`Overview` --> `Picture-in-Picture`
`Linked` --> `Side-by-side`

Classes:
`OverviewDetailViewer` --> `PictureInPictureViewer`
`LinkedDetailViewer` --> `SideBySideViewer`

Thoughts?  Does this work?

manzt replied:
Hmm. This is still confusing to me. There is a button that toggles between `hide/show picture-in-picture` and the same behavior is observed with the `overview` button when views are linked. I would prefer two toggles at the high level that are mutually exclusive :

`show/hide picture-in-picture` and `show/hide side-by-side`. If one of these buttons is on, the other is disabled. If `hide side-by-side` (so `SideBySideViewer` is in use), then conditionally render the lock pan / lock zoom buttons. If the user wants to try out the other view, they must first go through VivViewer. This will make it so the end-user doesn't expect the side-by-side view to also be able to show a picture-in-picture when in use.

ilan-gold commented:
@mccalluc I addressed most of your comments in the recent push.  The only two things are 1) I do think people can use non side-by-side by using overview/detail (and all the options within those two are configurable).  Does that answer your concerns about that?  As for the generated docs, I don't think I can style the markdown inline code (if that is what you mean) but I do think that there are customizations available with his package.  It seems a little out of scope, but it is [possible](https://github.com/documentationjs/documentation/blob/master/docs/THEMING.md).  Maybe we file an issue?

ilan-gold commented:
@manzt I think I also addressed most of what you had to say.  The last thing that you both seem to have commented on is naming.  I proposed something [here](https://github.com/hubmapconsortium/vitessce-image-viewer/pull/146#discussion_r408328975) but would be open to other suggestions.
